### PR TITLE
Add `Duration::from_nanos_u128`

### LINKED ISSUE #139201: Tracking Issue for `Duration::from_nanos_u128`
Feature gate: `#![feature(duration_from_nanos_u128)]`

This is a tracking issue for `Duration::from_nanos_u128`. See the ACP at https://github.com/rust-lang/libs-team/issues/567.

### Public API

```rust
// std::time

impl Duration {
    /// Creates a new Duration from the specified number of nanoseconds.
    ///
    /// # Panics
    ///
    /// Panics if the given number of nanoseconds overflows the Duration size.
    pub const fn from_nanos_u128(nanos: u128) -> Duration;
}
```

### Steps / History

- [X] Implementation: #145969
- [ ] Final comment period (FCP)[^1]
- [ ] Stabilization PR

### Unresolved Questions

- Is this the best name we can come up with?
- Should we instead try to [generalize `from_nanos`](https://github.com/rust-lang/rust/issues/139201#issuecomment-2925596603) with a sealed trait? That would block this on const traits.

[^1]: https://std-dev-guide.rust-lang.org/feature-lifecycle/stabilization.html

### PR DESCRIPTION
**Pull request by omanirudh**
Feature gate: `#![feature(duration_from_nanos_u128)]`
ACP: https://github.com/rust-lang/libs-team/issues/567
Tracking issue:  rust-lang/rust#139201

### CODE REVIEW

File: library/core/src/time.rs [hunk 1]
```diff
@@ -308,6 +308,29 @@ impl Duration {
...
+    /// which is around 584 years and above 584.94 years to be a little more precise. 
+    /// 
+    /// # Examples 
+    /// 
+    /// ```
+    /// use std::time::Duration;
+    /// let time_in_nanos = 2.pow(64);
+    /// let duration = Duration::from_nanos_u128(time_in_nanos);
+    /// 
+    /// ```
```
Comment (madhav-madhusoodanan) [inline library/core/src/time.rs:323]: You might want to add this stability attribute too.

```suggestion
    /// ```
    #[unstable(feature = "duration_from_nanos_u128", issue = "139201")]
    
```

File: library/core/src/time.rs [hunk 2]
```diff
@@ -308,6 +308,29 @@ impl Duration {
...
+    pub const fn from_nanos_u128(nanos: u128) -> Duration {
+        const NANOS_PER_SEC: u128 = self::NANOS_PER_SEC as u128;
+        let secs = nanos / NANOS_PER_SEC;
```
Comment (madhav-madhusoodanan) [inline library/core/src/time.rs:326]: Miri seems to have a [solution](https://github.com/rust-lang/miri/blob/cffbc13f5ce138dcbc13dcbbe682419ffa796a05/src/clock.rs#L49-L56) for handling cases when the number of seconds is greater than `u64::MAX`.

Maybe a better way to do this is (might need confirmation from someone who would know better):
```suggestion
        let secs = u64::try_from(duration / self::NANOS_PER_SEC).unwrap_or(u64::MAX);
```

Comment (Skgland) [reply library/core/src/time.rs:326]: The ACP and Tracking issue contain a doc commend specifying panic on overflow instead of saturating on overflow.
Also its a bit weird that on overflow the seconds saturate but the nanoseconds aren't saturated as well.

**Comment by RalfJung** [comment]
This needs a libs reviewer. I just proposed the API. :)

r? libs

**Comment by omanirudh** [comment]
> Thanks for taking this over! However, this needs a libs reviewer. I just proposed the API. :)
> 
> r? libs

okay

File: library/core/src/time.rs [hunk 3]
```diff
@@ -308,6 +308,30 @@ impl Duration {
...
+    /// which is around 584 years and above 584.94 years to be a little more precise. 
+    /// 
+    /// # Examples 
+    /// 
+    /// ```
+    /// use std::time::Duration;
+    /// let time_in_nanos = 2.pow(64);
+    /// let duration = Duration::from_nanos_u128(time_in_nanos);
+    /// 
+    /// ```
```
Comment (tgross35) [inline library/core/src/time.rs:323]: Needs the `#![feature(...)]` gate to work, also has a trailing newline.

**Review by tgross35 — CHANGES_REQUESTED** [review]

File: library/core/src/time.rs [hunk 4]
```diff
@@ -308,6 +308,30 @@ impl Duration {
...
+
+    /// Creates a new Duration from the specified number of nanoseconds.
+    /// 
+    /// Use this function if you need to specify time greater than what can fit in u64 
+    /// (around 584 years).
+    /// 
+    /// # Examples 
+    /// 
+    /// ```
+    /// use std::time::Duration;
```
Comment (madhav-madhusoodanan) [inline library/core/src/time.rs:320]: What @tgross35 meant was to add the feature gate `#![feature(duration_from_nanos_u128)]` _within_ this example.

Just for your understanding on why this is needed:
1. You're marking this function behind an `#[unstable(feature = ...)]` annotation. Think of it as the "lock" to this functionality.
2. Users can use this particular functionality only when they use the corresponding "key" in their codebase, which is the `#![feature(...)]` annotation.
3. This example in the comment is in reality a Rustdoc test which is used in the CI to test the usage of this functionality. So, to enable the testing for this functionality, you'll need to add the annotation _within_ this example.

So all you need to do is:

```suggestion
    /// #![feature(duration_from_nanos_u128)]
    /// use std::time::Duration;
```

Comment (omanirudh) [reply library/core/src/time.rs:320]: Hey thanks for the explanation, Will add this within the doc comment.

File: library/core/src/time.rs (hunk 3 above)
Comment (omanirudh) [reply library/core/src/time.rs]: Hey @tgross35  thanks for the review. 
will make changes and get get back.

**Comment by omanirudh** [comment]
@rustbot ready
r? @tgross35 
apologies for the many commits. too many errors.

File: library/core/src/time.rs [hunk 5]
```diff
@@ -308,6 +307,35 @@ impl Duration {
...
+    /// let time_in_nanos = 2u128.pow(64);
+    /// let duration = Duration::from_nanos_u128(time_in_nanos);
+    /// ```
+    #[unstable(feature = "duration_from_nanos_u128", issue = "139201")]
+    #[must_use]
+    #[inline]
+    pub const fn from_nanos_u128(nanos: u128) -> Duration {
+        const NANOS_PER_SEC: u128 = self::NANOS_PER_SEC as u128;
+        let secs: u64 = (nanos / NANOS_PER_SEC) as u64;
```
Comment (Skgland) [inline library/core/src/time.rs:331]: The function documentation currently states that the function panics on overflow,
but this cast truncates the result of the division to u64 instead.

e.g. `u128::MAX` results in `13088917067439035463.768211455s` instead of panicking.

[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=a7dd4f02019a6076d496aab832128686)

File: library/core/src/time.rs [hunk 6]
```diff
@@ -308,6 +307,38 @@ impl Duration {
         Duration { secs, nanos: subsec_nanos }
     }
 
+    /// Creates a new Duration from the specified number of nanoseconds.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the given number of weeks overflows the `Duration` size.
```
Comment (madhav-madhusoodanan) [inline library/core/src/time.rs:314]: Do mention the size beyond which this function would panic. 

Since `secs` is u64 and `nanos` is u32, the maximum size that can be fit is `(u64::MAX * NANOS_PER_SEC) + u32::MAX `.

A text like the following would be helpful:

```rust
/// Panics if the given number of nanoseconds is greater than what Duration can handle, 
/// which is `(u64::MAX * NANOS_PER_SEC) + u32::MAX`.
```

File: library/core/src/time.rs [hunk 7]
```diff
@@ -308,6 +307,38 @@ impl Duration {
...
+    #[must_use]
+    #[inline]
+    pub const fn from_nanos_u128(nanos: u128) -> Duration {
+        const NANOS_PER_SEC: u128 = self::NANOS_PER_SEC as u128;
+        let secs: u128 = (nanos / NANOS_PER_SEC);
+        if (secs > u64::MAX.into()) {
+            panic!("overflow in duration in Duration::from_nanos_u128");
+        }
+        let subsec_nanos = (nanos % NANOS_PER_SEC) as u32;
+        // SAFETY: x % 1_000_000_000 < 1_000_000_000
```
Comment (madhav-madhusoodanan) [inline library/core/src/time.rs:336]: `new_unchecked` is unsafe since there is no bounds check being performed on the input function. Here is the [reference](https://github.com/rust-lang/rust/blob/9b2e419449de861e51c376abd0642e1ff05fe531/library/core/src/num/niche_types.rs#L52).

(The comment saying that `val == 0 is language UB` doesn't exactly apply here since the `Nanosecond` type is [defined](https://github.com/rust-lang/rust/blob/9b2e419449de861e51c376abd0642e1ff05fe531/library/core/src/num/niche_types.rs#L106) to be between 0 and 999_999_999 both inclusive).

Just so that our safety assumption is clear here, do add the fact that `subsec_nanos >= 0` because `u128 >=0` and `u32 > 0`

**Comment by madhav-madhusoodanan** [comment]
Rest of the code looks good, amazing work!

Only need to make small changes on the documentation side of things (along with @Skgland 's review about panicking on overflow instead of truncating on overflow).

**Comment by madhav-madhusoodanan** [comment]
About the commit history (this is completely upto you)

You can use `git rebase -i HEAD~<number of commits>`. This opens up an interactive window using which you can reorganize and meld related commits together.

This is a related [thread](https://rust-lang.zulipchat.com/#narrow/channel/122651-general/topic/Good.20guide.20for.20.22how.20to.20squash.20a.20PR.22.3F/with/511675396) on zulip on how commits can be handled.

File: Cargo.lock [hunk 8]
```diff
@@ -225,7 +225,7 @@ dependencies = [
  "memchr",
  "serde",
  "serde_derive",
- "winnow 0.7.4",
```
Comment (madhav-madhusoodanan) [inline Cargo.lock:228]: If I may ask, why this change?

File: library/core/src/time.rs (hunk 6 above)
Comment (Skgland) [reply library/core/src/time.rs:314]: Shouldn't the last valid value be `(u64::MAX * NANOS_PER_SEC) + NANOS_PER_SEC - 1` as Duration doesn't use the full range of the `u32` for the nano seconds

Comment (madhav-madhusoodanan) [reply library/core/src/time.rs:314]: Ahh yes you're right, my mistake.

File: library/core/src/time.rs [hunk 9]
```diff
@@ -308,6 +307,39 @@ impl Duration {
         Duration { secs, nanos: subsec_nanos }
     }
 
+    /// Creates a new Duration from the specified number of nanoseconds.
```
Comment (tgross35) [inline library/core/src/time.rs:310]:
```suggestion
    /// Creates a new `Duration` from the specified number of nanoseconds.
```

File: library/core/src/time.rs [hunk 10]
```diff
@@ -308,6 +307,39 @@ impl Duration {
...
+    ///
+    /// # Panics
+    ///
+    /// Panics if the given number of nanoseconds is greater than what Duration can handle,
```
Comment (tgross35) [inline library/core/src/time.rs:314]:
```suggestion
    /// Panics if the given number of nanoseconds is greater than what `Duration` can handle,
```

File: library/core/src/time.rs [hunk 11]
```diff
@@ -308,6 +307,39 @@ impl Duration {
...
+    /// which is `(u64::MAX * NANOS_PER_SEC) + NANOS_PER_SEC - 1`
+    /// Use this function if you need to specify time greater than what can fit in u64
```
Comment (tgross35) [inline library/core/src/time.rs:316]:
```suggestion
    /// Use this function if you need to specify time greater than what can fit in `u64`
```

File: library/core/src/time.rs [hunk 12]
```diff
@@ -1,5 +1,4 @@
 #![stable(feature = "duration_core", since = "1.25.0")]
-
```
Comment (tgross35) [inline library/core/src/time.rs:2]: Looks like a spurious removal

**Review by tgross35 — CHANGES_REQUESTED** [review]
Implementation and docs look good to me, left a few small comments about adding backticks around code types. However, this needs some tests. Could you update `library/coretests/tests/time.rs` to include `from_nanos_u128` everywhere that `from_nanos` is tested? Also make sure some large values are tested, and one or two `#[should_panic]` tests for the overflow.

**Comment by tgross35** [comment]
@omanirudh are you able to do the updates to get this over the line? Mostly just test updates and a few nits.

**Comment by omanirudh** [comment]
Hey Trevor, thanks for bringing this to my attention.
I am caught up in my day job. I need time until Thursday morning (Eastern).
I would really love to do this and make my first OSS contribution.

Also I am tied up so if it doesn't happen by this time. I will move on and
close the pull request to keep it open for others.

All of this assuming there is time until then.
In case this isn't the case. Please feel free to assign this to
someone more immediately available.

**Comment by tgross35** [comment]
It's not a problem! This is open source, there are no deadlines 🙂 we just try to make sure changes that are pretty close to done don't fall off the map.

**Comment by omanirudh** [comment]
Awesome! Here's one cheer to make this happen soon.

**Comment by tgross35** [comment]
@omanirudh it looks like you wound up with bad commit history somehow. You should be able to fix this by rebasing your branch, check out https://rustc-dev-guide.rust-lang.org/git.html#rebasing

**Comment by omanirudh** [comment]
Re-triggering CI after rebase.
@rustbot

**Comment by tgross35** [comment]
It looks like you did a merge rather than a rebase - that will need to get fixed since this branch still contains some weird history. Were you able to try what was suggested in rustc-dev-guide (linked above)?

**Comment by omanirudh** [comment]
Hey Trevor
Yes I did do a rebase.

However CI wasn't getting triggered.
Core Test's
time.rs file had an additional #test
which I resolved and merged on github

The same thing I did in local and pushed to my branch but CI wasn't picking
it up.

It was stuck with "conflicts" in the file.

No worries I will take a look and rebase again if needed.

The other weird thing would be two commits
with cargo.lock both times my repo was 500+ commits behind upstream master.

I will squish these and keep just two commits
1) for the function
2) for the tests

**Comment by omanirudh** [comment]
Hey all.
I'm unable to manage the git merge conflict
the logic is simple and done.

I tried figuring out the git part in vain.
Please take this over.

I no longer can give time to this.

Thanks all for the assistance

File: library/core/src/time.rs [hunk 13]
```diff
@@ -308,6 +308,39 @@ impl Duration {
         Duration { secs, nanos: subsec_nanos }
     }
 
+    /// Creates a new Duration from the specified number of nanoseconds.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the given number of nanoseconds is greater than what Duration can handle,
+    /// which is `(u64::MAX * NANOS_PER_SEC) + NANOS_PER_SEC - 1`
```
Comment (tgross35) [inline library/core/src/time.rs:316]:
```suggestion
    /// Panics if the given number of nanoseconds is greater than [`Duration::MAX`].
```

Technically we say this can vary by platform, so we don't need to give specifics.

File: library/core/src/time.rs [hunk 14]
```diff
@@ -308,6 +308,39 @@ impl Duration {
...
+    /// Use this function if you need to specify time greater than what can fit in u64
+    /// (around 584 years).
```
Comment (tgross35) [inline library/core/src/time.rs:318]:
```suggestion
```

I think this is reasonably straightforward

File: library/core/src/time.rs [hunk 15]
```diff
@@ -308,6 +308,39 @@ impl Duration {
...
+    /// let duration = Duration::from_nanos_u128(time_in_nanos);
+    /// ```
+    #[unstable(feature = "duration_from_nanos_u128", issue = "139201")]
+    #[must_use]
+    #[inline]
+    pub const fn from_nanos_u128(nanos: u128) -> Duration {
+        const NANOS_PER_SEC: u128 = self::NANOS_PER_SEC as u128;
+        let secs: u128 = nanos / NANOS_PER_SEC;
+        if secs > u64::MAX as u128 {
+            panic!("overflow in duration in Duration::from_nanos_u128");
```
Comment (tgross35) [inline library/core/src/time.rs:335]:
```suggestion
            panic!("overflow in `Duration::from_nanos_u128`");
```

File: library/core/src/time.rs [hunk 16]
```diff
@@ -308,6 +308,39 @@ impl Duration {
...
+    /// ```
+    /// #![feature(duration_from_nanos_u128)]
+    /// use std::time::Duration;
+    /// let time_in_nanos = 2u128.pow(64);
+    /// let duration = Duration::from_nanos_u128(time_in_nanos);
+    /// ```
+    #[unstable(feature = "duration_from_nanos_u128", issue = "139201")]
+    #[must_use]
+    #[inline]
+    pub const fn from_nanos_u128(nanos: u128) -> Duration {
```
Comment (tgross35) [inline library/core/src/time.rs:331]:
```suggestion
    #[track_caller]
    pub const fn from_nanos_u128(nanos: u128) -> Duration {
```

Makes the panic message show where this was called, rather than reporting std's `time.rs` as the panic location

File: library/coretests/tests/time.rs
Comment (tgross35) [inline library/coretests/tests/time.rs:1]: You'll need to add `#![feature(duration_from_nanos_u128)]` to `coretests/tests/lib.rs`, even for testing the unstable feature needs to be enabled (this is your CI failure)

File: library/coretests/tests/time.rs [hunk 17]
```diff
@@ -123,6 +136,8 @@ fn nanos() {
     assert_eq!(Duration::from_micros(1_000_001).subsec_nanos(), 1000);
     assert_eq!(Duration::from_nanos(999_999_999).subsec_nanos(), 999_999_999);
     assert_eq!(Duration::from_nanos(1_000_000_001).subsec_nanos(), 1);
+    assert_eq!(Duration::from_nanos_u128(999_999_999).subsec_nanos(), 999_999_999);
+    assert_eq!(Duration::from_nanos_u128(1_000_000_001).subsec_nanos(), 1);
 }
```
Comment (tgross35) [inline library/coretests/tests/time.rs:142]: Please also add a test to `duration_const` below (look for `const NANOS`)

**Review by tgross35 — CHANGES_REQUESTED** [review]
